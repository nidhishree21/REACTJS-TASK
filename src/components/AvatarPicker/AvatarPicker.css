@import url("https://fonts.googleapis.com/css?family=Source+Sans+Pro");

.TitleText {
  font-family: "Source Sans Pro", sans-serif;
  font-size: 16px;
  text-align: center;
  color: rgb(249, 249, 249);
  margin-top: 14px;
  margin-bottom: 14px;
}
.ActiveAvatarWrapper {
  max-width: 62px;
  margin-left: auto;
  margin-right: auto;
}

.ModalWrapper {
  max-width: 280px;
  margin-left: auto;
  margin-right: auto;
}
