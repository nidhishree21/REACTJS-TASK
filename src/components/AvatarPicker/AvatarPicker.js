import React, { Component } from "react";
import PropTypes from "prop-types";
import Modal from "./Modal/Modal";
import Avatar from "./Avatar/Avatar";
import AvatarList from "./AvatarList/AvatarList";
import networkReq from "../../shared/networkReq";
import styles from "./AvatarPicker.css";

class AvatarPicker extends Component {
  static propTypes = {
    avatars: PropTypes.arrayOf(PropTypes.object).isRequired
  };

  state = {
    activeAvatar: this.props.avatars[0],
    showModal: false,
    isUpdatingAvatar: false,
    updatingAvatarID: -1
  };

  onClickActiveAvatar = e => {
    e.preventDefault();
    if (e.keyCode === 13 || e.keyCode === 32 || e.type === "click") {
      this.setState(prevState => {
        return { showModal: true };
      });
    }
  };

  onCloseModal = e => {
    e.preventDefault();
    if (e.keyCode === 27 || e.type === "click") {
      this.setState(prevState => {
        return { showModal: false, updatingAvatarID: -1 };
      });
      document.removeEventListener("keyup", this.onCloseModal, false);
    }
  };

  onNetworkResponse = avatar => {
    this.setState(prevState => {
      return {
        activeAvatar: avatar,
        showModal: false,
        isUpdatingAvatar: false,
        updatingAvatarID: -1
      };
    });
    document.removeEventListener("keyup", this.onCloseModal, false);
  };

  callNetwork = async payload => {
    const response = await networkReq(payload);
    return await this.onNetworkResponse(response);
  };

  onClickListAvatar = async e => {
    e.preventDefault();
    const targetAvatarID = +e.target.id;
    if (e.keyCode === 13 || e.keyCode === 32 || e.type === "click") {
      this.setState(prevState => ({
        updatingAvatarID: targetAvatarID,
        isUpdatingAvatar: true
      }));
      return await this.callNetwork({ avatarID: targetAvatarID });
    }
  };

  highlight = avatarID => {
    return avatarID === this.state.activeAvatar.id;
  };

  showSpinner = avatarID => {
    return avatarID === this.state.updatingAvatarID;
  };

  render() {
    return (
      <div className={styles.AvatarPickerWrapper}>
        <div className={styles.ActiveAvatarWrapper}>
          <Avatar
            avatar={this.state.activeAvatar}
            onClickAvatar={this.onClickActiveAvatar}
            isActive={true}
            highlight={false}
            showSpinner={false}
          />
        </div>
        <div className={styles.ModalWrapper}>
          <Modal show={this.state.showModal} hide={this.onCloseModal}>
            <p className={styles.TitleText}>Choose your avatar</p>
            <AvatarList
              avatars={this.props.avatars}
              onClickAvatar={this.onClickListAvatar}
              isActive={false}
              highlight={this.highlight}
              showSpinner={this.showSpinner}
            />
          </Modal>
        </div>
      </div>
    );
  }
}

export default AvatarPicker;
