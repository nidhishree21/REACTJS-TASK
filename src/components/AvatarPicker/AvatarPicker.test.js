import React from "react";
import { shallow } from "enzyme";
import AvatarPicker from "./AvatarPicker";

const testProps = {
  avatars: [
    {
      src: "/static/media/avatar1.7467b519.png",
      label: "Avatar 1",
      id: 1
    },
    {
      src: "/static/media/avatar3.c1a8b069.png",
      label: "Avatar 2",
      id: 2
    }
  ]
};

describe("AvatarPicker", () => {
  const wrapper = shallow(<AvatarPicker {...testProps} />);
  it("renders AvatarPicker correctly", () => {
    expect(wrapper).toMatchSnapshot("Initial Render");
  });
});

describe("onClickListAvatar", () => {
  const wrapper = shallow(<AvatarPicker {...testProps} />);

  it("renders the view with updated Active Avatar", async () => {
    // before vs after - change in active Avatar props.
    expect(wrapper).toMatchSnapshot("1. Before clicking an avatar in the list");
    await wrapper.instance().onClickListAvatar({
      type: "click",
      target: { id: 2 },
      preventDefault: jest.fn(() => {})
    });
    wrapper.update();
    expect(wrapper).toMatchSnapshot("2. After clicking an avatar in the list");
  });
});

describe("onClickActiveAvatar", () => {
  const wrapper = shallow(<AvatarPicker {...testProps} />);
  // before vs after - change in Modal props.
  it("renders the view with Modal", async () => {
    expect(wrapper).toMatchSnapshot("1. Before clicking active avatar");
    await wrapper.instance().onClickActiveAvatar({
      type: "click",
      preventDefault: jest.fn(() => {})
    });
    wrapper.update();
    expect(wrapper).toMatchSnapshot("2. After clicking active avatar");
  });
});

describe("onCloseModal", () => {
  const wrapper = shallow(<AvatarPicker {...testProps} />);
  // before vs after - change in Modal props.
  it("hides Modal from the view", async () => {
    //display Modal first
    await wrapper.instance().onClickActiveAvatar({
      type: "click",
      preventDefault: jest.fn(() => {})
    });
    wrapper.update();
    expect(wrapper).toMatchSnapshot("1. Before hiding Modal");
    //hide Modal
    await wrapper.instance().onCloseModal({
      type: "click",
      preventDefault: jest.fn(() => {})
    });
    wrapper.update();
    expect(wrapper).toMatchSnapshot("2. After hiding Modal");
  });
});
