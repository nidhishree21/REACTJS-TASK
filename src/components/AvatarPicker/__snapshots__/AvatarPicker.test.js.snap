// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. Before clicking active avatar 1`] = `
<div
  className="AvatarPickerWrapper"
>
  <div
    className="ActiveAvatarWrapper"
  >
    <Avatar
      avatar={
        Object {
          "id": 1,
          "label": "Avatar 1",
          "src": "/static/media/avatar1.7467b519.png",
        }
      }
      highlight={false}
      isActive={true}
      onClickAvatar={[Function]}
      showSpinner={false}
    />
  </div>
  <div
    className="ModalWrapper"
  >
    <Modal
      hide={[Function]}
      show={false}
    >
      <p
        className="TitleText"
      >
        Choose your avatar
      </p>
      <AvatarList
        avatars={
          Array [
            Object {
              "id": 1,
              "label": "Avatar 1",
              "src": "/static/media/avatar1.7467b519.png",
            },
            Object {
              "id": 2,
              "label": "Avatar 2",
              "src": "/static/media/avatar3.c1a8b069.png",
            },
          ]
        }
        highlight={[Function]}
        isActive={false}
        onClickAvatar={[Function]}
        showSpinner={[Function]}
      />
    </Modal>
  </div>
</div>
`;

exports[`1. Before clicking an avatar in the list 1`] = `
<div
  className="AvatarPickerWrapper"
>
  <div
    className="ActiveAvatarWrapper"
  >
    <Avatar
      avatar={
        Object {
          "id": 1,
          "label": "Avatar 1",
          "src": "/static/media/avatar1.7467b519.png",
        }
      }
      highlight={false}
      isActive={true}
      onClickAvatar={[Function]}
      showSpinner={false}
    />
  </div>
  <div
    className="ModalWrapper"
  >
    <Modal
      hide={[Function]}
      show={false}
    >
      <p
        className="TitleText"
      >
        Choose your avatar
      </p>
      <AvatarList
        avatars={
          Array [
            Object {
              "id": 1,
              "label": "Avatar 1",
              "src": "/static/media/avatar1.7467b519.png",
            },
            Object {
              "id": 2,
              "label": "Avatar 2",
              "src": "/static/media/avatar3.c1a8b069.png",
            },
          ]
        }
        highlight={[Function]}
        isActive={false}
        onClickAvatar={[Function]}
        showSpinner={[Function]}
      />
    </Modal>
  </div>
</div>
`;

exports[`1. Before hiding Modal 1`] = `
<div
  className="AvatarPickerWrapper"
>
  <div
    className="ActiveAvatarWrapper"
  >
    <Avatar
      avatar={
        Object {
          "id": 1,
          "label": "Avatar 1",
          "src": "/static/media/avatar1.7467b519.png",
        }
      }
      highlight={false}
      isActive={true}
      onClickAvatar={[Function]}
      showSpinner={false}
    />
  </div>
  <div
    className="ModalWrapper"
  >
    <Modal
      hide={[Function]}
      show={true}
    >
      <p
        className="TitleText"
      >
        Choose your avatar
      </p>
      <AvatarList
        avatars={
          Array [
            Object {
              "id": 1,
              "label": "Avatar 1",
              "src": "/static/media/avatar1.7467b519.png",
            },
            Object {
              "id": 2,
              "label": "Avatar 2",
              "src": "/static/media/avatar3.c1a8b069.png",
            },
          ]
        }
        highlight={[Function]}
        isActive={false}
        onClickAvatar={[Function]}
        showSpinner={[Function]}
      />
    </Modal>
  </div>
</div>
`;

exports[`2. After clicking active avatar 1`] = `
<div
  className="AvatarPickerWrapper"
>
  <div
    className="ActiveAvatarWrapper"
  >
    <Avatar
      avatar={
        Object {
          "id": 1,
          "label": "Avatar 1",
          "src": "/static/media/avatar1.7467b519.png",
        }
      }
      highlight={false}
      isActive={true}
      onClickAvatar={[Function]}
      showSpinner={false}
    />
  </div>
  <div
    className="ModalWrapper"
  >
    <Modal
      hide={[Function]}
      show={true}
    >
      <p
        className="TitleText"
      >
        Choose your avatar
      </p>
      <AvatarList
        avatars={
          Array [
            Object {
              "id": 1,
              "label": "Avatar 1",
              "src": "/static/media/avatar1.7467b519.png",
            },
            Object {
              "id": 2,
              "label": "Avatar 2",
              "src": "/static/media/avatar3.c1a8b069.png",
            },
          ]
        }
        highlight={[Function]}
        isActive={false}
        onClickAvatar={[Function]}
        showSpinner={[Function]}
      />
    </Modal>
  </div>
</div>
`;

exports[`2. After clicking an avatar in the list 1`] = `
<div
  className="AvatarPickerWrapper"
>
  <div
    className="ActiveAvatarWrapper"
  >
    <Avatar
      avatar={
        Object {
          "id": 2,
          "label": "Avatar 2",
          "src": "test-file-stub",
        }
      }
      highlight={false}
      isActive={true}
      onClickAvatar={[Function]}
      showSpinner={false}
    />
  </div>
  <div
    className="ModalWrapper"
  >
    <Modal
      hide={[Function]}
      show={false}
    >
      <p
        className="TitleText"
      >
        Choose your avatar
      </p>
      <AvatarList
        avatars={
          Array [
            Object {
              "id": 1,
              "label": "Avatar 1",
              "src": "/static/media/avatar1.7467b519.png",
            },
            Object {
              "id": 2,
              "label": "Avatar 2",
              "src": "/static/media/avatar3.c1a8b069.png",
            },
          ]
        }
        highlight={[Function]}
        isActive={false}
        onClickAvatar={[Function]}
        showSpinner={[Function]}
      />
    </Modal>
  </div>
</div>
`;

exports[`2. After hiding Modal 1`] = `
<div
  className="AvatarPickerWrapper"
>
  <div
    className="ActiveAvatarWrapper"
  >
    <Avatar
      avatar={
        Object {
          "id": 1,
          "label": "Avatar 1",
          "src": "/static/media/avatar1.7467b519.png",
        }
      }
      highlight={false}
      isActive={true}
      onClickAvatar={[Function]}
      showSpinner={false}
    />
  </div>
  <div
    className="ModalWrapper"
  >
    <Modal
      hide={[Function]}
      show={false}
    >
      <p
        className="TitleText"
      >
        Choose your avatar
      </p>
      <AvatarList
        avatars={
          Array [
            Object {
              "id": 1,
              "label": "Avatar 1",
              "src": "/static/media/avatar1.7467b519.png",
            },
            Object {
              "id": 2,
              "label": "Avatar 2",
              "src": "/static/media/avatar3.c1a8b069.png",
            },
          ]
        }
        highlight={[Function]}
        isActive={false}
        onClickAvatar={[Function]}
        showSpinner={[Function]}
      />
    </Modal>
  </div>
</div>
`;

exports[`Initial Render 1`] = `
<div
  className="AvatarPickerWrapper"
>
  <div
    className="ActiveAvatarWrapper"
  >
    <Avatar
      avatar={
        Object {
          "id": 1,
          "label": "Avatar 1",
          "src": "/static/media/avatar1.7467b519.png",
        }
      }
      highlight={false}
      isActive={true}
      onClickAvatar={[Function]}
      showSpinner={false}
    />
  </div>
  <div
    className="ModalWrapper"
  >
    <Modal
      hide={[Function]}
      show={false}
    >
      <p
        className="TitleText"
      >
        Choose your avatar
      </p>
      <AvatarList
        avatars={
          Array [
            Object {
              "id": 1,
              "label": "Avatar 1",
              "src": "/static/media/avatar1.7467b519.png",
            },
            Object {
              "id": 2,
              "label": "Avatar 2",
              "src": "/static/media/avatar3.c1a8b069.png",
            },
          ]
        }
        highlight={[Function]}
        isActive={false}
        onClickAvatar={[Function]}
        showSpinner={[Function]}
      />
    </Modal>
  </div>
</div>
`;
